TypeScript issues in v3.1.0

## 🐛 Bug Report

With xstyled@3.1.0 and emotion, SystemProps does not infer correctly. 

For example, `py={4}` makes TypeScript error `Type '4' is not assignable to type 'SystemProp<PaddingTop<0 | (string & {})> & PaddingBottom<0 | (string & {})>, Theme> | undefined'` with [creating a d.ts file](https://xstyled.dev/docs/typescript/).

Without custom d.ts, `pt={y}` is valid, but wrong utilities are also valid such as `opacity={{ xxx: 'typo' }}`.

## To Reproduce

- [Install latest xstyled with emotion](https://xstyled.dev/docs/installation/#using-emotion)
- Create styled.d.ts and paste [the example d.ts](https://xstyled.dev/docs/typescript/#create-a-declaration-file-1)
- Pass numeric value to system props, e.g. `<x.div py={4}>`
- `yarn tsc --noEmit`

```
src/pages/index.tsx:12:7 - error TS2322: Type '4' is not assignable to type 'SystemProp<PaddingTop<0 | (string & {})> & PaddingBottom<0 | (string & {})>, Theme> | undefined'.

12       py={4}
         ~~

  node_modules/@xstyled/system/dist/index.d.ts:600:5
    600     py?: SystemProp<ThemeSpace<T> | (CSS.Property.PaddingTop & CSS.Property.PaddingBottom), T>;
            ~~
    The expected type comes from property 'py' which is declared here on type 'IntrinsicAttributes & SystemProps<Theme> & { as?: ElementType<any> | undefined; theme?: Theme | undefined; } & Omit<...> & { ...; }'
```

## Expected behavior

With/without d.ts,

- `<x.div mt={4}>` must not be type error
- `<x.div opacity={{ hov: 'typo' }}>` must be type error

## Link to repl or repo (highly encouraged)

prepared two examples to reproduce.

- without d.ts https://github.com/yamitzky/template-next.js-xstyled/blob/repro/3.1.0-wo-style/src/pages/index.tsx
- with [d.ts](https://xstyled.dev/docs/typescript/#use-typescript-with-emotion) https://github.com/yamitzky/template-next.js-xstyled/blob/repro/3.1.0-with-custom-style/src/pages/index.tsx


## Run `npx envinfo --system --binaries --npmPackages @xstyled/system,@xstyled/styled-components,styled-components --markdown --clipboard`

Paste the results here:

```bash
## System:
 - OS: macOS 11.4
 - CPU: (8) arm64 Apple M1
 - Memory: 87.31 MB / 8.00 GB
 - Shell: 5.8 - /bin/zsh
## Binaries:
 - Node: 16.7.0 - ~/.nvm/versions/node/v16.7.0/bin/node
 - Yarn: 1.22.10 - /opt/homebrew/bin/yarn
 - npm: 7.20.3 - ~/.nvm/versions/node/v16.7.0/bin/npm
## npmPackages:
 - @emotion/react: ^11.4.1 => 11.4.1 
 - @emotion/styled: ^11.3.0 => 11.3.0 
 - @xstyled/emotion: ^3.1.0 => 3.1.0 
 - @xstyled/system: ^3.1.0 => 3.1.0 
```

--- a/packages/system/src/types.ts
+++ b/packages/system/src/types.ts
@@ -120,7 +120,7 @@ declare type SynthesizedPath<T extends {}> = {
     | (T[P] extends { [key: string]: any; [key: number]: any }
         ? `${string & P}` | `${string & P}.${SynthesizedPath<T[P]>}`
         : `${string & P}`)
-    | number
+    | (number & P)
 }[T extends any[] ? number & keyof T : keyof T]
 
 export type ThemeNamespaceValue<

--- a/packages/system/src/types.ts
+++ b/packages/system/src/types.ts
@@ -71,7 +71,8 @@ export type ThemeStates<T extends ITheme> = T extends {
   ? T['states']
   : unknown
 
-export type ThemeVariants<T extends ITheme> = ThemeScreens<T> & Omit<ThemeStates<T>, '_'>
+export type ThemeVariants<T extends ITheme> = ThemeScreens<T> &
+  Omit<ThemeStates<T>, '_'>
 
 export type ThemeProp<TType, TTheme extends ITheme> = {
   [P in keyof ThemeVariants<TTheme>]?: TType | ThemeProp<TType, TTheme>
@@ -115,9 +116,11 @@ export interface TransformValue {
 }
 
 declare type SynthesizedPath<T extends {}> = {
-  [P in keyof T]: (T[P] extends {[key: string]: any; [key: number]: any}
-    ? `${string & P}` | `${string & P}.${SynthesizedPath<T[P]>}`
-    : `${string & P}`) | number
+  [P in keyof T]:
+    | (T[P] extends { [key: string]: any; [key: number]: any }
+        ? `${string & P}` | `${string & P}.${SynthesizedPath<T[P]>}`
+        : `${string & P}`)
+    | number
 }[T extends any[] ? number & keyof T : keyof T]
 
 export type ThemeNamespaceValue<

